File: apps/macos/Tests/CueWeaveMacTests/LocalSettingsTests.swift
import Foundation
import Testing
@testable import CueWeaveMac

@Suite("Local settings")
struct LocalSettingsTests {
    @Test("API keys use only an owner-readable local file")
    func roundTripAndPermissions() throws {
        let directory = FileManager.default.temporaryDirectory
            .appendingPathComponent("cueweave-settings-\(UUID().uuidString)")
        let url = directory.appendingPathComponent("config.json")
        defer { try? FileManager.default.removeItem(at: directory) }
        let expected = LocalSettings(
            alignmentProvider: "ai_studio",
            openRouterAPIKey: "openrouter-secret",
            aiStudioAPIKey: "aistudio-secret",
            openRouterModel: ModelDefaults.openRouter,
            aiStudioModel: ModelDefaults.aiStudio,
            uiLanguage: "zh"
        )

        try LocalSettingsStore.save(expected, to: url)

        #expect(LocalSettingsStore.load(from: url) == expected)
        #expect(permissions(url) == 0o600)
        #expect(permissions(directory) == 0o700)
    }

    @Test("Legacy defaults migrate without replacing custom models")
    func migrateLegacyDefaults() throws {
        let directory = FileManager.default.temporaryDirectory
            .appendingPathComponent("cueweave-settings-migration-\(UUID().uuidString)")
        let legacyURL = directory.appendingPathComponent("legacy.json")
        let customURL = directory.appendingPathComponent("custom.json")
        defer { try? FileManager.default.removeItem(at: directory) }

        try LocalSettingsStore.save(LocalSettings(
            openRouterModel: "google/gemini-3.7-flash",
            aiStudioModel: "gemini-3.7-flash"
        ), to: legacyURL)
        try LocalSettingsStore.save(LocalSettings(
            openRouterModel: "custom/openrouter-model",
            aiStudioModel: "custom-aistudio-model"
        ), to: customURL)

        let migrated = LocalSettingsStore.load(from: legacyURL)
        #expect(migrated.openRouterModel == ModelDefaults.openRouter)
        #expect(migrated.aiStudioModel == ModelDefaults.aiStudio)
        let custom = LocalSettingsStore.load(from: customURL)
        #expect(custom.openRouterModel == "custom/openrouter-model")
        #expect(custom.aiStudioModel == "custom-aistudio-model")
    }

    private func permissions(_ url: URL) -> Int {
        let attributes = try? FileManager.default.attributesOfItem(atPath: url.path)
        return attributes?[.posixPermissions] as? Int ?? 0
    }
}

File: apps/windows/CueWeave.Windows.Tests/LocalSettingsTests.cs
using CueWeave.WinUI.Models;
using CueWeave.WinUI.Services;

namespace CueWeave.Windows.Tests;

[TestClass]
public sealed class LocalSettingsTests
{
    [TestMethod]
    public void SettingsRoundTripUsesPlainLocalJson()
    {
        var path = Path.Combine(Path.GetTempPath(), "cueweave-settings-test", Guid.NewGuid().ToString("N"), "settings.json");
        try
        {
            LocalSettingsStore.Save(new LocalSettings {
                AlignmentProvider = "ai_studio",
                OpenRouterApiKey = "openrouter-secret",
                AiStudioApiKey = "aistudio-secret",
                UiLanguage = "zh",
            }, path);
            var loaded = LocalSettingsStore.Load(path);
            Assert.AreEqual("ai_studio", loaded.AlignmentProvider);
            Assert.AreEqual("openrouter-secret", loaded.OpenRouterApiKey);
            Assert.AreEqual("aistudio-secret", loaded.AiStudioApiKey);
            Assert.AreEqual(ModelDefaults.OpenRouter, loaded.OpenRouterModel);
            Assert.AreEqual(ModelDefaults.AiStudio, loaded.AiStudioModel);
            Assert.AreEqual("zh", loaded.UiLanguage);
            StringAssert.Contains(LocalSettingsStore.ConfigPath, "CueWeave");
            StringAssert.Contains(LocalSettingsStore.ConfigPath, "settings.json");
        }
        finally
        {
            var root = Directory.GetParent(Path.GetDirectoryName(path)!)?.FullName;
            if (root is not null && Directory.Exists(root)) Directory.Delete(root, true);
        }
    }

    [TestMethod]
    public void LegacyDefaultsMigrateWithoutReplacingCustomModels()
    {
        var directory = Path.Combine(Path.GetTempPath(), "cueweave-settings-test", Guid.NewGuid().ToString("N"));
        var legacyPath = Path.Combine(directory, "legacy.json");
        var customPath = Path.Combine(directory, "custom.json");
        try
        {
            LocalSettingsStore.Save(new LocalSettings {
                OpenRouterModel = "google/gemini-3.7-flash",
                AiStudioModel = "gemini-3.7-flash",
            }, legacyPath);
            LocalSettingsStore.Save(new LocalSettings {
                OpenRouterModel = "custom/openrouter-model",
                AiStudioModel = "custom-aistudio-model",
            }, customPath);

            var migrated = LocalSettingsStore.Load(legacyPath);
            Assert.AreEqual(ModelDefaults.OpenRouter, migrated.OpenRouterModel);
            Assert.AreEqual(ModelDefaults.AiStudio, migrated.AiStudioModel);
            var custom = LocalSettingsStore.Load(customPath);
            Assert.AreEqual("custom/openrouter-model", custom.OpenRouterModel);
            Assert.AreEqual("custom-aistudio-model", custom.AiStudioModel);
        }
        finally
        {
            if (Directory.Exists(directory)) Directory.Delete(directory, true);
        }
    }
}
